### DISCUSSION TITLE
Can post a portable USB MIDI implementation for S3 (the one in Tulip seems nice) that functions with Arduino IDE? (to receive MIDI from a DAW on laptop through S3 USB)

### DISCUSSION THREAD

**Discussion by 01GOD** [shorepine/tulipcc#323] [General]
Hi! Thanks for posting Tulip! 
Can post a portable USB MIDI implementation for S3 (seems Brian Whitman likely knows how to port that from Tulip easily) that functions with Arduino IDE? (to receive MIDI from a DAW on laptop through S3 USB)

I have been working on that for hours today and it's obvious (from many searches and some code testing today) that the ESP32 synth community needs that. 

Thanks in advance, if possible! 

PS: Yes, of course in C/C++.

**Comment by bwhitman** [comment]
Surprising that this doesn't exist! I'm sure it's not too hard to get our `usb_keyboard.c` / `midi.c` MIDI implementation into Arduino. I'll add it to my list of things to check out.

**Reply by 01GOD** [reply]
Saw now. Thanks in advance!

**Reply by 01GOD** [reply]
Yes, it was a surprising number of hours work yesterday searching and sifting through code and trying different things. The "Arduino MIDI Library" (https://github.com/FortySevenEffects/arduino_midi_library) ReadMe says it now features USB MIDI, but there wasn't any example that I found after installed that library in the Arduino IDE. 

I have done some MIDI in Unity or iOS/iPadOS before (a long while ago though) and really would prefer to go with community DRY principles on this rather than making such a thing again. 

By the way, THANKFULLY finished designing (and confirming functionality of) an ESP32-S3 PCB recently and likely adding a cap touch display and a DAC connector to it soon, so that may be a good alternative to put Tulip on. Was hoping to make an SBC with a cap touchscreen to run android on soon, but I should seemingly do this first. If recommend any recent open hardware android or linux boards, that is appreciated also.

**Comment by dpwe** [comment]
Our implementation relies on the espressif USB library, which is nontrivial and I think makes specific assumptions about the USB hardware choices on the ESP32.  There are probably lots of USB libraries for Arduino, and lots of hardware variants that have to be supported.  I don't think porting our code will be easy because of the way it assumes espressif's API.

**Reply by RocketManRC** [reply]
This is easy to accomplish if you use PlatformIO instead of the Arduino IDE. You can still use the Arduino Framework just not the IDE so your project will be compatible with all the other available Arduino libraries. I have used this approach a lot and can cobble together a simple example if there is interest. You can even have USB MIDI and USB Serial at the same time.

The reason for using PlatformIO for this is to be able to set the proper build flags and integrate the necessary libraries. Maybe it can be done with the Arduino IDE as well but I don't use it anymore so I can't offer any help with that.

**Reply by 01GOD** [reply]
Ah. I had asked the person that made AcidBox (https://github.com/copych/AcidBox) about PlatformIO and that person stated that there were some issues with it when they tried it, and they said Arduino IDE 2.x+ is based on VSCode, so I haven't tried PlatformIO properly yet. 

It's likely good to be familiar with how to use all 3 though (and there is maybe still some mild IDE snobbery toward Arduino IDE, LOL), so I may have a proper go at PlatformIO (and maybe Expressif IDE) this week. I'm learning PCB design and doing casing design also, so juggling a bunch of things. 

Anyway, will be great to have a "THIS DEFINITELY FUNCTIONS" USB MIDI from laptop DAW to ESP32-S3 (and other compatible ESP32 boards) example soon. I did get this other example to function for sending MIDI notes from an ESP32-S3 (tinyUSB style): https://github.com/esp32beans/ESP32USBMIDI/blob/main/examples/ESP32USBMIDI/ESP32USBMIDI.ino

I will likely look at if it's possible to do bidirectional MIDI with that code somehow today.

**Comment by bwhitman** [comment]
Perhaps there's some confusion - are we asking about turning a ESP32 board into a MIDI peripheral (that is, you plug it into a computer and it's recognized as a USB MIDI port that you can send notes to), or allow a ESP32 board to have a MIDI USB peripheral plugged into it (for example, a USB->MIDI adapter, or keyboard or synth with USB MIDI)? Tulip only does the latter.

**Reply by 01GOD** [reply]
Basically to run a synth on an ESP32 and have it be controlled through USB MIDI by a DAW on a laptop. I suppose from the look of the Tulip promos, it looks like it may be running as a USB host to communicate with USB MIDI keyboards mainly.

**Reply by bwhitman** [reply]
Yes, Tulip is meant as a USB host for midi and typing keyboards. 

It would be possible to port our USB host implementation into Arduino, but it sounds like that's not what you're looking for.

Maybe over on [AMY](https://github.com/shorepine/amy) we can discuss a better ESP32-S3 Arduino example that also supports a TinyUSB backed MIDI USB endpoint.

**Reply by 01GOD** [reply]
Sounds good! Should I start a discussion on the Amy repo with a similar title to this thread? 
It's past 1am here, so I should get to bed. Can talk about it more tomorrow. Thanks in advance for assisting with making a functional Arduino IDE based ESP32-S3 solution for USB MIDI for the ESP32 synth community.
